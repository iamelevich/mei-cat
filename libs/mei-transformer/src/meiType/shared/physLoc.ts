import { Type } from "typebox";
import { StandardTagSchema } from "../common";
import { AttrBiblSchema } from "./attr/bibl";
import { AttrCommonSchema } from "./attr/common";
import { AttrFacsimileSchema } from "./attr/facsimile";

/**
 * Groups information about the current physical location of a bibliographic item, such as the repository in which it is located and its shelf mark(s), and its previous locations.
 * @see https://music-encoding.org/guidelines/v5/elements/physLoc.html
 */
export const PhysLocSchema = Type.Intersect([
	StandardTagSchema,
	AttrCommonSchema,
	AttrBiblSchema,
	AttrFacsimileSchema,
	Type.Object(
		{
			// Content model according to MEI spec:
			// rng:zeroOrMore - model.headLike
			/**
			 * Contains any heading, for example, the title of a section of text, or the heading of a list.
			 * @see https://music-encoding.org/guidelines/v5/elements/head.html
			 */
			head: Type.Optional(Type.Union([Type.Ref("head"), Type.Array(Type.Ref("head"))])),

			// rng:zeroOrMore - group of model.repositoryLike and model.identifierLike
			/**
			 * Institution, agency, or individual which holds a bibliographic item.
			 * @see https://music-encoding.org/guidelines/v5/elements/repository.html
			 */
			repository: Type.Optional(
				Type.Union([Type.Ref("repository"), Type.Array(Type.Ref("repository"))]),
			),

			/**
			 * An alpha-numeric string that establishes the identity of the described material.
			 * @see https://music-encoding.org/guidelines/v5/elements/identifier.html
			 */
			identifier: Type.Optional(
				Type.Union([Type.Ref("identifier"), Type.Array(Type.Ref("identifier"))]),
			),

			// rng:optional - history
			/**
			 * Provides a container for information about the history of a resource other than the circumstances of its creation.
			 * @see https://music-encoding.org/guidelines/v5/elements/history.html
			 */
			history: Type.Optional(Type.Ref("history")),
		},
		{ additionalProperties: false },
	),
]);
