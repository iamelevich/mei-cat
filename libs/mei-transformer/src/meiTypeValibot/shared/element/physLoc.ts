import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import {
	AttrBiblSchema,
	AttrCommonSchema,
	ModelHeadLikeSchema,
	ModelIdentifierLikeSchema,
	ModelRepositoryLikeSchema,
} from "..";
import { AttrFacsimileSchema } from "../../facsimile";
import { HistorySchema } from "../../header";

/**
 * Base schema with attribute, to simplify types for PhysLocSchema
 */
const PhysLocBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrFacsimileSchema.entries,
});

/**
 * Groups information about the current physical location of a bibliographic item, such as the repository in which it is located and its shelf mark(s), and its previous locations.
 * @see https://music-encoding.org/guidelines/v5/elements/physLoc.html
 */
export const PhysLocSchema = v.intersect([
	PhysLocBaseSchema,
	v.object({
		/**
		 * Reference to element history
		 * @see https://music-encoding.org/guidelines/v5/elements/history.html
		 */
		history: v.optional(v.lazy(() => HistorySchema)),
	}),
	ModelHeadLikeSchema,
	ModelIdentifierLikeSchema,
	ModelRepositoryLikeSchema,
]);

export type PhysLocData = v.InferOutput<typeof PhysLocSchema>;
