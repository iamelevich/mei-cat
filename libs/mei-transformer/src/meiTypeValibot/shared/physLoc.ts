import * as v from "valibot";
import { StandardTagSchema } from "../common";
import { AttrCommonSchema } from "./attr/common";
import { AttrLangSchema } from "./attr/lang";

/**
 * Simplified schema for physLoc element.
 * @see https://music-encoding.org/guidelines/v5/elements/physLoc.html
 */
export const PhysLocSchema = v.object({
	...StandardTagSchema.entries,
	...AttrCommonSchema.entries,
	...AttrLangSchema.entries,
});

export type PhysLocData = v.InferOutput<typeof PhysLocSchema>;
