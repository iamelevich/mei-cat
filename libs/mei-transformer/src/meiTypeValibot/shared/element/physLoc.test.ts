import { describe, expect, it } from "bun:test";
import * as v from "valibot";
import { PhysLocSchema } from "./physLoc";

describe("PhysLocSchema", () => {
	it("should be defined", () => {
		expect(PhysLocSchema).toBeDefined();
	});

	it("should parse a physLoc with repository and identifier", () => {
		const parseResult = v.parse(PhysLocSchema, {
			repository: {
				identifier: {
					"#text": "DK-Kk",
					"@auth.uri": "http://www.rism.info/",
					"@auth": "RISM",
					"@xml:id": "identifier_N21072",
				},
				"@xml:id": "repository_N21071",
			},
			identifier: {
				"#text": "CNA IAc",
				"@xml:id": "identifier_N21076",
			},
			"@xml:id": "physLoc_N21070",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a physLoc with just repository", () => {
		const parseResult = v.parse(PhysLocSchema, {
			repository: {
				identifier: {
					"#text": "DK-Kk",
					"@auth": "RISM",
					"@xml:id": "identifier_N21072",
				},
				"@xml:id": "repository_N21071",
			},
			"@xml:id": "physLoc_N21070",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a physLoc with just identifier", () => {
		const parseResult = v.parse(PhysLocSchema, {
			identifier: {
				"#text": "CNA IAc",
				"@xml:id": "identifier_N21076",
			},
			"@xml:id": "physLoc_N21070",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a minimal physLoc", () => {
		const parseResult = v.parse(PhysLocSchema, {
			"@xml:id": "physLoc_N21070",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a physLoc with multiple identifiers", () => {
		const parseResult = v.parse(PhysLocSchema, {
			identifier: [
				{
					"#text": "CNA IAc",
					"@xml:id": "identifier_N21076",
				},
				{
					"#text": "NKS 5155, 4Â°",
					"@xml:id": "identifier_N2109D",
				},
			],
			"@xml:id": "physLoc_N21070",
		});

		expect(parseResult).toMatchSnapshot();
	});
});
