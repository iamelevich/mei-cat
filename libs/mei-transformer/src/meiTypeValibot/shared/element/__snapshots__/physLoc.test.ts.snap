// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`PhysLocSchema should parse a physLoc with repository and identifier 1`] = `
{
  "@xml:id": "physLoc_N21070",
  "identifier": {
    "#text": "CNA IAc",
    "@xml:id": "identifier_N21076",
  },
  "repository": {
    "@xml:id": "repository_N21071",
    "identifier": {
      "#text": "DK-Kk",
      "@auth": "RISM",
      "@auth.uri": "http://www.rism.info/",
      "@xml:id": "identifier_N21072",
    },
  },
}
`;

exports[`PhysLocSchema should parse a physLoc with just repository 1`] = `
{
  "@xml:id": "physLoc_N21070",
  "repository": {
    "@xml:id": "repository_N21071",
    "identifier": {
      "#text": "DK-Kk",
      "@auth": "RISM",
      "@xml:id": "identifier_N21072",
    },
  },
}
`;

exports[`PhysLocSchema should parse a physLoc with just identifier 1`] = `
{
  "@xml:id": "physLoc_N21070",
  "identifier": {
    "#text": "CNA IAc",
    "@xml:id": "identifier_N21076",
  },
}
`;

exports[`PhysLocSchema should parse a minimal physLoc 1`] = `
{
  "@xml:id": "physLoc_N21070",
}
`;

exports[`PhysLocSchema should parse a physLoc with multiple identifiers 1`] = `
{
  "@xml:id": "physLoc_N21070",
  "identifier": [
    {
      "#text": "CNA IAc",
      "@xml:id": "identifier_N21076",
    },
    {
      "#text": "NKS 5155, 4Â°",
      "@xml:id": "identifier_N2109D",
    },
  ],
}
`;
